File: app/src/main/java/viewmodel/CategoryViewModel.java
package viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

import model.Category;
import repo.dbRepoController.CategoryDBRepository;

public class CategoryViewModel extends ViewModel {

    private final CategoryDBRepository dbRepository;
    private final MutableLiveData<List<Category>> categoriesLiveDate;

    private final MutableLiveData<Boolean> goToTopLiveData;

    public CategoryViewModel() {
        dbRepository = new CategoryDBRepository();
        categoriesLiveDate = dbRepository.getCategoriesLiveData();
        goToTopLiveData = new MutableLiveData<>();
    }

    public void fetch() {
        dbRepository.fetchAllCategories();
    }

    public List<Category> getAllCategories() {
        try {
            return dbRepository.getAllCategories();
        } catch (InterruptedException e) {
            return null;
        }

    }

    public void addCategory(Category category) {
        try {
            dbRepository.addCategory(category);
        } catch (InterruptedException ignored) {
        }
    }

    public void editCategory(Category category) {
        try {
            dbRepository.editCategory(category);
        } catch (InterruptedException ignored) {
        }
    }

    public void deleteCategory(Category category) {
        try {
            dbRepository.deleteCategory(category);
        } catch (InterruptedException ignored) {
        }
    }

    public void deleteAllCategories() {
        try {
            dbRepository.deleteAllCategories();
        } catch (InterruptedException ignored) {
        }
    }

    public long getCategoriesCount() {
        try {
            return dbRepository.categoriesCount();
        } catch (InterruptedException e) {
            return 0;
        }
    }

    public boolean categoriesIsEmpty() {
        return getCategoriesCount() == 0;
    }

    public boolean pureValidateCategory(Category category) {
        String categoryName = category.getName();
        return categoryName != null && !categoryName.trim().isEmpty() && categoryName.trim().length() != 0;
    }

    public String validateCategory(Category category) {
        boolean isValidCategory = pureValidateCategory(category);

        if (!isValidCategory)
            return "عنوان دسته نمی تواند خالی باشد!";

        return null;
    }

    public LiveData<List<Category>> getCategoriesLiveDate() {
        return categoriesLiveDate;
    }

    public void goToTop() {
        if (goToTopLiveData != null)
            goToTopLiveData.setValue(true);
    }

    public LiveData<Boolean> getGoToTopLiveData() {
        return goToTopLiveData;
    }
}

File: app/src/main/java/viewmodel/CategoryViewModel.kt
package viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import model.Category
import repo.dbRepoController.CategoryDBRepository
import javax.inject.Inject

@HiltViewModel
class CategoryViewModel @Inject constructor(
    private val repo: CategoryDBRepository
) : ViewModel() {

    val categoriesFlow: SharedFlow<List<Category>?> = repo.getAllCategories().shareIn(
        viewModelScope,
        SharingStarted.WhileSubscribed(),
        1
    )

    //--------------channels----------------
    private val _goToTopChannel: Channel<Boolean> = Channel()
    val goToTopFlow: Flow<Boolean> = _goToTopChannel.receiveAsFlow()

    suspend fun getCategoriesCount(): Long {
        var count = 0L

        viewModelScope.launch(Dispatchers.IO) {
            count = repo.categoriesCount()
        }.join()

        return count
    }

    suspend fun categoriesIsEmpty(): Boolean {
        var isEmpty = false

        viewModelScope.launch(Dispatchers.IO) {
            isEmpty = repo.categoriesCount() == 0L
        }.join()

        return isEmpty
    }

    fun addCategory(category: Category?) {
        viewModelScope.launch(Dispatchers.IO) {
            repo.addCategory(category)
        }
    }

    fun editCategory(category: Category?) {
        viewModelScope.launch(Dispatchers.IO) {
            repo.editCategory(category!!)
        }
    }

    fun deleteCategory(category: Category?) {
        viewModelScope.launch(Dispatchers.IO) {
            repo.deleteCategory(category!!)
        }
    }

    fun deleteAllCategories() {
        viewModelScope.launch(Dispatchers.IO) {
            repo.deleteAllCategories()
        }
    }

    private fun pureValidateCategory(category: Category): Boolean {
        val categoryName = category.name
        return categoryName != null &&
                categoryName.trim().isNotEmpty() &&
                categoryName.trim().isNotEmpty()
    }

    fun validateCategory(category: Category): String? {
        val isValidCategory = pureValidateCategory(category)
        return if (!isValidCategory) "عنوان دسته نمی تواند خالی باشد!" else null
    }

    fun goToTop() {
        viewModelScope.launch {
            _goToTopChannel.send(true)
        }
    }

}